### PR TITLE
Total confidence: wip

### PR DESCRIPTION
**Pull request by nickybondarenko**
#### :heavy_check_mark: Checklist

- [x ] All tests are passing
- [ ] Relevant documentation updated
- [x ] linter/style run on changed files
- [x ] Tests added for new functionality
- [ ] Regression tests added for bug fixes
- [ ] Tested in a corresponding example app

### CODE REVIEW

File: packages/sdk/src/Confidence.test.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      expect(confidence.getContext()).toEqual(newContext);
+    });
+  });
+  describe('withContext', () => {
+    it('creates a child context', () => {
+      const parent = new Confidence({} as any);
+      const additionalContext = {
+        clothes: 'pants',
+      };
+      let child = parent.withContext(additionalContext);
```
Comment (nicklasl) [inline packages/sdk/src/Confidence.test.ts:39]:
```suggestion
      const child = parent.withContext(additionalContext);
```

File: packages/sdk/src/Confidence.test.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      expect(child.getContext()).toEqual({});
+      expect(parent.getContext()).toEqual({
+        clothes: 'pants',
+      });
+      expect(Object.keys(child.getContext())).toEqual([]);
+      child.clearContext()
+      expect(child.getContext()).toEqual({
+        clothes: 'pants',
+      });
+    })
```
Comment (nicklasl) [inline packages/sdk/src/Confidence.test.ts:71]: nice test 👍

File: packages/sdk/src/Confidence.test.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    it('creates a new confidence object', () => {
+      const confidenceProvider = Confidence.create({
+        clientSecret: 'secret',
+        region: 'us',
+        baseUrl: 'https://www.spotify.com',
+        environment: 'client',
+        fetchImplementation: {} as any,
+        timeout: 10,
+      });
+      expect(confidenceProvider.getContext()).toEqual({});
```
Comment (nicklasl) [inline packages/sdk/src/Confidence.test.ts:83]: weird name of const.
```suggestion
      const confidence = Confidence.create({
        clientSecret: 'secret',
        region: 'us',
        baseUrl: 'https://www.spotify.com',
        environment: 'client',
        fetchImplementation: {} as any,
        timeout: 10,
      });
      expect(confidence.getContext()).toEqual({});
```

File: packages/sdk/src/Confidence.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
+import { FlagResolverClient, FlagResolution, ApplyManager } from './FlagResolverClient';
+import { EventSenderEngine } from './EventSenderEngine';
+import { Value } from './Value';
+import { EventSender } from './events';
+import { Context } from './context';
+
+export { FlagResolverClient, FlagResolution };
```
Comment (nicklasl) [inline packages/sdk/src/Confidence.ts:7]: What does it mean that we export these types? Should we do that already?

File: packages/sdk/src/context.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  getContext(): Context;
+  setContext(context: Context): void;
+  updateContextEntry<K extends string>(name: K, value: Context[K]): void;
+  removeContextEntry(name: string): void;
+  withContext(context: Context): Self;
+  clearContext(): void;
+}
+
+export interface Context extends Value.Struct {
+  openFeature?: Value.Struct & {
```
Comment (nicklasl) [inline packages/sdk/src/context.ts:13]: Not blocking at all for this initial PR but just as a note: 
We want to declare any semantic context keys here asap in order to risk collisions later on (collisions such as customers defining context entries that we later plan as semantic contexts).

File: packages/sdk/src/events.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { Value } from './Value';
+import { Contextual } from './context';
+
+export interface EventSender extends Contextual<EventSender> {
+  sendEvent(name: string, message?: Value): void;
```
Comment (nicklasl) [inline packages/sdk/src/events.ts:5]: message should be a Struct?

File: packages/sdk/src/EventSenderEngine.ts [hunk 7]
```diff
@@ -1,8 +1,116 @@
...
+    } else {
+      if (this.flushTimeoutMilliseconds > 0) {
+        this.pendingFlush = setTimeout(() => {
+          this.flush();
+        }, this.flushTimeoutMilliseconds);
+      }
+    }
+  }
+
+  public flush(): Promise<boolean> {
```
Comment (nicklasl) [inline packages/sdk/src/EventSenderEngine.ts:94]: what does it mean that this is function is `public` while `send` is not?

File: packages/sdk/src/EventSenderEngine.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      eventTime: new Date().toISOString(),
+      payload: { ...message, ...context },
+    });
+    if (this.pendingFlush) {
+      clearTimeout(this.pendingFlush);
+    }
+    if (this.writeQueue.length >= this.maxBatchSize) {
+      this.flush();
+    } else {
+      if (this.flushTimeoutMilliseconds > 0) {
```
Comment (nicklasl) [inline packages/sdk/src/EventSenderEngine.ts:86]: What does it mean when `flushTimeoutMilliseconds` <= 0?
That we only flush on batch size?

File: packages/sdk/src/EventSenderEngine.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        return request;
+      })
+      .build(fetchImplementation);
+  }
+
+  send(context: Value.Struct, name: string, message?: Value.Struct): void {
+    this.writeQueue.push({
+      eventDefinition: `eventDefinitions/${name}`,
+      eventTime: new Date().toISOString(),
```
Comment (nicklasl) [inline packages/sdk/src/EventSenderEngine.ts:77]: nit: Worth sending a clock interface to be able to unit test the time things?

File: packages/sdk/src/EventSenderEngine.ts (hunk 8 above)
Comment (nickybondarenko) [reply packages/sdk/src/EventSenderEngine.ts:88]: In this case, all negative and 0 should mean that we don't set a timeout and just flush when a batch is full.

File: packages/sdk/src/EventSenderEngine.ts (hunk 9 above)
Comment (nickybondarenko) [reply packages/sdk/src/EventSenderEngine.ts:79]: We can use fake timers to do that

**Review by nicklasl — APPROVED** [review]

**Pull request merged**